File: 05-behavioral-patterns/01-cor/01-cor.go
package main

import "fmt"

type Handler interface {
	Handle(request map[string]string) string
	SetNext(handler Handler)
}

type BaseHandler struct {
	next Handler
}

func (b *BaseHandler) SetNext(handler Handler) {
	b.next = handler
}

func (b *BaseHandler) Handle(request map[string]string) string {
	if b.next != nil {
		return b.next.Handle(request)
	}
	return "No handler could process the request"
}

type AuthHandler struct {
	BaseHandler
}

func (a *AuthHandler) Handle(request map[string]string) string {
	if request["user"] == "admin" {
		return "Authenticated"
	}
	return a.BaseHandler.Handle(request)
}

type LoggingHandler struct {
	BaseHandler
}

func (l *LoggingHandler) Handle(request map[string]string) string {
	fmt.Println("Logging request:", request)
	return l.BaseHandler.Handle(request)
}

func main() {
	req := map[string]string{"user": "guest"}

	log := &LoggingHandler{}
	auth := &AuthHandler{}

	log.SetNext(auth)

	fmt.Println(log.Handle(req))
}

File: 05-behavioral-patterns/01-cor/01-cor.py
class Handler:
    def __init__(self, next_handler=None):
        self.next = next_handler

    def handle(self, request):
        if self.next:
            return self.next.handle(request)
        return "No handler could process the request"


class AuthHandler(Handler):
    def handle(self, request):
        if request.get("user") == "admin":
            return "Authenticated"
        return super().handle(request)


class LoggingHandler(Handler):
    def handle(self, request):
        print(f"Logging request: {request}")
        return super().handle(request)


request = {"user": "guest"}
chain = LoggingHandler(AuthHandler())

result = chain.handle(request)
print(result)
